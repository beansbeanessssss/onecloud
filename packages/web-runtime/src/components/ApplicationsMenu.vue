<template>
  <nav :aria-label="$gettext('Main navigation')" class="uk-flex uk-flex-middle">
    <oc-button
      id="_appSwitcherButton"
      ref="menubutton"
      v-oc-tooltip="applicationSwitcherLabel"
      appearance="raw"
      variation="inverse"
      class="oc-topbar-menu-burger"
      :aria-label="applicationSwitcherLabel"
    >
      <oc-icon name="grid" size="large" class="uk-flex" />
    </oc-button>
    <oc-drop
      ref="menu"
      drop-id="app-switcher-dropdown"
      toggle="#_appSwitcherButton"
      mode="click"
      :options="{ pos: 'bottom-right', delayHide: 0 }"
      class="uk-width-large"
      padding-size="small"
      close-on-click
    >
      <ul class="oc-p-xs">
        <li
          v-for="(n, nid) in menuItems"
          :key="`apps-menu-${nid}`"
          style="list-style: none"
          class="oc-pt-s oc-pb-s"
        >
          <a
            v-if="n.url"
            key="apps-menu-external-link"
            :target="n.target"
            :href="n.url"
            style="display: inline-flex; vertical-align: top; gap: 10px"
          >
            <oc-icon :name="n.iconMaterial" size="medium" />
            <span v-text="$gettext(n.title)" />
          </a>
          <router-link
            v-else
            key="apps-menu-internal-link"
            :to="n.path"
            style="display: inline-flex; vertical-align: top; gap: 10px"
          >
            <oc-icon :name="n.iconMaterial" size="medium" />
            <span v-text="$gettext(n.title)" />
          </router-link>
        </li>
      </ul>
    </oc-drop>
  </nav>
</template>

<script>
import NavigationMixin from '../mixins/navigationMixin'
import UiKit from 'uikit'

export default {
  mixins: [NavigationMixin],
  props: {
    applicationsList: {
      type: Array,
      required: false,
      default: () => []
    }
  },
  computed: {
    menuItems() {
      return this.navigation_getMenuItems([null, 'apps', 'appSwitcher'])
    },
    applicationSwitcherLabel() {
      return this.$gettext('Application Switcher')
    }
  },
  mounted() {
    UiKit.util.on('#app-switcher-dropdown', 'shown', () => {
      this.focusFirstLink()
    })

    UiKit.util.on('#app-switcher-dropdown', 'hidden', () => {
      this.$emit('closed')
      this.focusMenuButton()
    })
  },
  methods: {
    logout() {
      this.visible = false
      this.$store.dispatch('logout')
    },
    focusFirstLink() {
      /*
       * Delay for two reasons:
       * - for screen readers Virtual buffer
       * - to outsmart uikit's focus management
       */
      setTimeout(() => {
        this.$refs.menu.$el.querySelector('a:first-of-type').focus()
      }, 500)
    },
    focusMenuButton() {
      this.$refs.menubutton.$el.focus()
    }
  }
}
</script>
