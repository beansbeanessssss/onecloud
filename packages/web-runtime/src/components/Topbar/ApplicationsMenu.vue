<template>
  <nav
    id="applications-menu"
    :aria-label="$gettext('Main navigation')"
    class="oc-flex oc-flex-middle"
  >
    <oc-button
      id="_appSwitcherButton"
      ref="menubutton"
      v-oc-tooltip="applicationSwitcherLabel"
      appearance="raw-inverse"
      variation="brand"
      class="oc-topbar-menu-burger"
      :aria-label="applicationSwitcherLabel"
    >
      <oc-icon name="grid" size="large" class="oc-flex" />
    </oc-button>
    <oc-drop
      ref="menu"
      drop-id="app-switcher-dropdown"
      toggle="#_appSwitcherButton"
      mode="click"
      padding-size="small"
      close-on-click
    >
      <oc-list class="applications-list">
        <li v-for="(n, nid) in applicationsList" :key="`apps-menu-${nid}`" @click="clickApp(n)">
          <oc-button
            :key="n.url ? 'apps-menu-external-link' : 'apps-menu-internal-link'"
            :type="n.url ? 'a' : 'router-link'"
            :target="n.target"
            :href="n.url"
            :to="n.path"
            :appearance="n.active ? 'raw-inverse' : 'raw'"
            :variation="n.active ? 'primary' : 'passive'"
            :class="{ 'oc-background-primary-gradient router-link-active': n.active }"
          >
            <span class="icon-box">
              <oc-icon :name="n.icon" variation="inherit" />
            </span>
            <span v-text="$gettext(n.title)" />
            <oc-icon v-if="n.active" name="check" class="active-check" variation="inherit" />
          </oc-button>
        </li>
      </oc-list>
    </oc-drop>
  </nav>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { configurationManager } from 'web-pkg/src/configuration'
import { urlJoin } from 'web-client/src/utils'

export default defineComponent({
  props: {
    applicationsList: {
      type: Array as PropType<any[]>,
      required: false,
      default: () => []
    }
  },
  computed: {
    applicationSwitcherLabel() {
      return this.$gettext('Application Switcher')
    }
  },
  mounted() {
    this.$refs.menu?.tippy?.setProps({
      onHidden: () => this.$refs.menubutton.$el.focus(),
      onShown: () => this.$refs.menu.$el.querySelector('a:first-of-type').focus()
    })
  },
  methods: {
    async clickApp(appEntry) {
      // @TODO use id or similar
      if (appEntry.url?.endsWith('/apps/files')) {
        await this.setClassicUIDefault()
      }
    },
    setClassicUIDefault() {
      const url = urlJoin(configurationManager.serverUrl, '/index.php/apps/web/settings/default')
      return this.$clientService.httpAuthenticated.post(url, { isDefault: false })
    }
  }
})
</script>

<style lang="scss" scoped>
.applications-list li {
  margin: var(--oc-space-xsmall) 0;

  a,
  button {
    gap: var(--oc-space-medium);
    justify-content: flex-start;
    width: 100%;

    &:focus,
    &:hover {
      background-color: var(--oc-color-background-hover);
      text-decoration: none;
    }

    .icon-box {
      display: inline-flex;
      justify-content: center;
      align-items: center;
      width: 40px;
      height: 40px;
    }

    .active-check {
      position: absolute;
      right: 1rem;
    }
  }
}
</style>
