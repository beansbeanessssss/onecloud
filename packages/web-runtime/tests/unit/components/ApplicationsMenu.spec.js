import { shallowMount, createLocalVue } from '@vue/test-utils'
import DesignSystem from 'owncloud-design-system'
import GetText from 'vue-gettext'

import ApplicationsMenu from 'web-runtime/src/components/ApplicationsMenu.vue'

const localVue = createLocalVue()
localVue.use(DesignSystem)
localVue.use(GetText, {
  translations: 'does-not-matter',
  silent: true
})

const menuLinks = [
  {
    name: 'Files',
    id: 'files',
    icon: 'folder',
    type: 'extension'
  },
  {
    name: 'External',
    id: 'external',
    icon: 'some-icon',
    type: '',
    url: 'http://some.org'
  }
]

describe('ApplicationsMenu component', () => {
  jest
    .spyOn(ApplicationsMenu.mixins[0].methods, 'navigation_getMenuItems')
    .mockImplementation(() => [
      {
        icon: 'folder',
        iconUrl: undefined,
        title: 'Files',
        path: '/files'
      },
      {
        icon: 'some-icon',
        iconUrl: undefined,
        title: 'External',
        url: 'http://some.org',
        target: '_blank'
      }
    ])

  it('should render navigation with button and menu items in dropdown', () => {
    const wrapper = getWrapper(menuLinks)
    expect(wrapper).toMatchSnapshot()
  })
})

function getWrapper(applicationsList = []) {
  return shallowMount(ApplicationsMenu, {
    localVue,
    propsData: {
      applicationsList
    },
    stubs: {
      'oc-button': true,
      'oc-icon': true,
      'oc-drop': true,
      'router-link': true
    },
    directives: {
      'oc-tooltip': jest.fn()
    },
    mocks: {
      $route: {},
      $router: {
        resolve: () => {
          return {
            href: 'href'
          }
        }
      }
    }
  })
}
