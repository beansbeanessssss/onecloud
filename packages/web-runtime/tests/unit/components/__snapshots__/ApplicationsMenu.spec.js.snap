// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationsMenu component shoud set aria-label prop on switcher button 1`] = `
<oc-button-stub type="button" size="medium" arialabel="Application Switcher" variation="inverse" appearance="raw" justifycontent="center" gapsize="medium" id="_appSwitcherButton" class="oc-topbar-menu-burger">
  <oc-icon-stub name="grid" filltype="fill" accessiblelabel="" type="span" size="large" variation="passive" class="uk-flex"></oc-icon-stub>
</oc-button-stub>
`;

exports[`ApplicationsMenu component should show app menus 1`] = `
<oc-drop-stub dropid="app-switcher-dropdown" toggle="#_appSwitcherButton" position="bottom-start" mode="click" closeonclick="true" paddingsize="small" options="[object Object]" class="uk-width-large">
  <ul class="oc-p-xs">
    <li class="oc-pt-s oc-pb-s" style="list-style: none;">
      <router-link-stub to="/files" tag="a" ariacurrentvalue="page" event="click" style="display: inline-flex; vertical-align: top; gap: 10px;">
        <oc-icon-stub name="folder" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub> <span>Files</span>
      </router-link-stub>
    </li>
    <li class="oc-pt-s oc-pb-s" style="list-style: none;"><a target="_blank" href="http://some.org" style="display: inline-flex; vertical-align: top; gap: 10px;">
        <oc-icon-stub name="some-icon" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub> <span>External</span>
      </a></li>
  </ul>
</oc-drop-stub>
`;
